### PR TITLE
fix errors when a sweep with skip failures solves once or no times

### LINKED ISSUE #700: unexplained bad value + no-valid-sweeps errors
@bqpd this is the issue I discussed with you today at the meeting. I'm running cvxopt and Windows 10. The file information is below

on the d8 github branch recl, file name is Flight Envelope.ipynb
commit #: 7e4d0c9d1e0e82b55df5c3f1dbf0428cd38a0114

with these slider settings 
image

I get this output:
```
Cost
----
 [ 0.67      0.67      ]

Sweep Variables
---------------
   T : [ 230       217      ]  [K]       Temperature at X-plane cruise        
 \mu : [ 1.49e-05  1.42e-05 ]  [kg/m/s]  Kinematic viscosity at X-plane cruise
\rho : [ 0.466     0.265    ]  [kg/m**3] Air density at X-plane cruise        
   h : [ 9e+03     1.3e+04  ]  [m]       Altitude                             

Free Variables
--------------
     (L/D)_X : [ 16.6      16.6     ]         Lift-to-drag ratio, X-plane                             
        CD_X : [ 0.0375    0.0375   ]         Cruise drag coefficient, X-plane                        
        CL_X : [ 0.623     0.623    ]         Cruise lift coefficent, X-plane                         
         M_X : [ 0.78      0.78     ]         Cruise Mach number, X-plane                             
        Re_X : [ 1.89e+07  1.89e+07 ]         Cruise Reynolds number, X-plane                         
           S : [ 0.67      0.67     ]         slack in slack variables                                
         S_X : [ 2.62e+03  2.62e+03 ]  [m**2] X-plane wing area                                       
V_{CL, a, g} : [ 41        54.3     ]  [kt]   CL-matching velocity (no ballast) with extra load factor
V_{CL, b, g} : [ 143       190      ]  [kt]   CL-matching velocity (ballast) with extra load factor   
    V_{CL,a} : [ 29        38.4     ]  [kt]   CL-matching velocity (no ballast)                       
    V_{CL,b} : [ 101       134      ]  [kt]   CL-matching velocity (ballast)                          
      V_{Ma} : [ 461       447      ]  [kt]   Mach-matching velocity                                  
      V_{Re} : [ 76.1      127      ]  [kt]   Re-matching velocity                                    
     V_{max} : [ 767       1.35e+03 ]  [kt]   Max-IAS velocity                                        
    V_{st,a} : [ 14.5      19.2     ]  [kt]   stall speed (no ballast)                                
    V_{st,b} : [ 50.6      67.1     ]  [kt]   stall speed (ballast)                                   
         W_X : [ 2.33e+05  2.33e+05 ]  [lbf]  X-plane weight                                          
           a : [ 304       295      ]  [m/s]  X-plane speed of sound                                  
      h_{ft} : [ 2.95e+04  4.27e+04 ]  [ft]   Altitude (ft)                                           
           n : [ 4.76      4.76     ]         X-plane scale factor    
```

n is a scale factor that should never go over 1....so 4.76 is obviously incorrect. Phillipe gets ~.5 which is the correct answer.

Furthermore, these sliders
image

lead to this error:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-f206d415f349> in make_plot(W0_val, Wball_val, CLD8_val, ReD8_val, ERC, WSmax_val, sRe_val, sCL_val, sLD_val, g_x_val)
    147     m = Model(objective, constraints, substitutions)
    148 
--> 149     sol = m.solve(verbosity=1,skipsweepfailures=True)
    150     pltAlt   = mag(sol('h_{ft}'))
    151     pltVst_a = mag(sol('V_{st,a}'))

c:\python27\lib\site-packages\gpkit\model.pyc in solve(self, solver, verbosity, skipsweepfailures, *args, **kwargs)
    340         try:
    341             return self._solve("gp", solver, verbosity, skipsweepfailures,
--> 342                                *args, **kwargs)
    343         except ValueError as err:
    344             if err.message == ("GeometricPrograms cannot contain Signomials"):

c:\python27\lib\site-packages\gpkit\model.pyc in _solve(self, programType, solver, verbosity, skipsweepfailures, *args, **kwargs)
    485                                          " To ignore such failures, solve with"
    486                                          " skipsweepfailures=True.")
--> 487             for var, val in solution["constants"].items():
    488                 solution["constants"][var] = [val[0]]
    489 

KeyError: 'constants'
```

which is also something I don't think Philippe is seeing

Finally, I also get this error

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-f206d415f349> in make_plot(W0_val, Wball_val, CLD8_val, ReD8_val, ERC, WSmax_val, sRe_val, sCL_val, sLD_val, g_x_val)
    147     m = Model(objective, constraints, substitutions)
    148 
--> 149     sol = m.solve(verbosity=1,skipsweepfailures=True)
    150     pltAlt   = mag(sol('h_{ft}'))
    151     pltVst_a = mag(sol('V_{st,a}'))

c:\python27\lib\site-packages\gpkit\model.pyc in solve(self, solver, verbosity, skipsweepfailures, *args, **kwargs)
    340         try:
    341             return self._solve("gp", solver, verbosity, skipsweepfailures,
--> 342                                *args, **kwargs)
    343         except ValueError as err:
    344             if err.message == ("GeometricPrograms cannot contain Signomials"):

c:\python27\lib\site-packages\gpkit\model.pyc in _solve(self, programType, solver, verbosity, skipsweepfailures, *args, **kwargs)
    505         self.solution = solution  # NOTE: SIDE EFFECTS
    506         if verbosity > 0:
--> 507             print(solution.table())
    508         return solution
    509 

c:\python27\lib\site-packages\gpkit\solution_array.pyc in table(self, tables, fixedcols, included_models, excluded_models)
    106             else:
    107                 strs += [" %-.4g" % self["cost"]]
--> 108                 cost_units = self.program.cost.units
    109             strs[-1] += unitstr(cost_units, into=" [%s] ", dimless="")
    110             strs += [""]

AttributeError: 'list' object has no attribute 'cost'

In [ ]:
```

​This one will flash up during the solve and before the KeyError, I'm not sure the end state of a solve has ever been this error...

### LINKED ISSUE #707: constants KeyError when skipsweepfailure=True
The latest code on the `recl` branch of the D8 repo gives the following error:

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-6-7f68fcb57527> in make_plot(Wball_val, CLD8_val, ReD8_val, ERC, WSmax_val, sRe_val, sCL_val, sLD_val, g_x_val, cost_val)
    156     m = Model(objective, constraints, substitutions)
    157 
--> 158     sol = m.solve(verbosity=1,skipsweepfailures=True)
    159     print sol('n')
    160     pltAlt   = mag(sol('h_{ft}'))

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/constraints/prog_factories.pyc in solvefn(self, solver, verbosity, skipsweepfailures, *args, **kwargs)
     75             run_sweep(genfunction, self, solution, skipsweepfailures,
     76                       constants, sweep, linkedsweep,
---> 77                       solver, verbosity-1, *args, **kwargs)
     78         else:
     79             self.program, solvefn = genfunction(self, verbosity-1)

/Users/philippekirschen/Documents/MIT/Research/GPkit/gpkit/gpkit/constraints/prog_factories.pyc in run_sweep(genfunction, self, solution, skipsweepfailures, constants, sweep, linkedsweep, solver, verbosity, *args, **kwargs)
    139     solution["sweepvariables"] = KeyDict()
    140     ksweep, klinkedsweep = KeyDict(sweep), KeyDict(linkedsweep)
--> 141     for var, val in solution["constants"].items():
    142         if var in ksweep or var in klinkedsweep:
    143             solution["sweepvariables"][var] = val

KeyError: 'constants'
```

Any ideas @bqpd ?

### PR DESCRIPTION
**Pull request by bqpd**
Closes #707, closes #700, ready for review.

### CODE REVIEW

**Comment by mayork** [comment]
@whoburg or @bqpd could this be merged so we have working code for a meeting with Aurora tomorrow morning?

**Comment by bqpd** [comment]
...this doesn't actually fix your issue, I believe? Also, you can check out this branch separately.

**Comment by mayork** [comment]
@bqpd ah sorry didn't realize this was on a different branch..this solves this key attribute error but now I get  AttributeError: 'list' object has no attribute 'cost'

this is the same error I fully describe at the bottom of https://github.com/hoburg/gpkit/issues/700

sorry for the bad naming of the issuse...Philippe and I thought it was a windows thing since he wasn't getting the same error, but now we think I had pulled gpkit more recently than him

**Comment by bqpd** [comment]
This PR should fix both #707 and #700. However the bigger issue, (see my comments on #707), is that almost _none_ of your sweeps solve!

**Comment by bqpd** [comment]
`'list' object has no attribute 'cost'` occurs on master when only one solve is successful out of many

the key attribute error occurs when none are.

**Comment by mayork** [comment]
okay that makes sense, I didn't realize the 'list' object has no attribute 'cost' error was due to the sweeps not solving, sorry for the confusion!

**Comment by bqpd** [comment]
No worries!

**Comment by bqpd** [comment]
todo: add a test

**Comment by whoburg** [comment]
fix before merging this: it appears this doesn't work on `gas_male_fixCDRhtry.py` at commit a1c6e23e5d43178bafa7571c575f264e41be9b94 in the `1682master` branch

```python
Using solver 'mosek'
Solving for 1451 variables.
Solving took 0.228 seconds.
Traceback (most recent call last):
  File "gas_male_fixCDRhtry.py", line 37, in <module>
    sol = M.solve('mosek')
  File "/Users/whoburg/MIT/dev/gpkit/gpkit/constraints/prog_factories.py", line 86, in solvefn
    print(solution.table())
  File "/Users/whoburg/MIT/dev/gpkit/gpkit/solution_array.py", line 139, in table
    self.program[0].cost).units
TypeError: 'GeometricProgram' object does not support indexing
```

**Comment by bqpd** [comment]
Oh, that's the most recent commit's fault. Knew I should have tested it.

**Comment by bqpd** [comment]
@galbramc, this seems to be not catching the `# pylint: disable=unsubscriptable-object` flag on line 142?

**Comment by bqpd** [comment]
retest this please

**Comment by bqpd** [comment]
@whoburg please merge this if it passes your review, it's an issue in Next Release

**Comment by whoburg** [comment]
retest this please

**Comment by whoburg** [comment]
wasn't there a plan to add a test?

File: gpkit/solution_array.py [hunk 1]
```diff
@@ -123,6 +123,7 @@ def table(self, tables=("cost", "sweepvariables", "freevariables",
             subdict = self.get(table, None)
             table_title = self.table_titles[table]
             if table == "cost":
+                # pylint: disable=unsubscriptable-object
```
Comment (bqpd) [inline gpkit/solution_array.py]: whoa, much cleaner.

**Comment by bqpd** [comment]
Whoops, yes.

File: gpkit/tests/t_sub.py [hunk 2]
```diff
@@ -150,6 +150,19 @@ def test_united_sub_sweep(self):
         firstcost = m.solve(verbosity=0)["cost"][0]
         self.assertAlmostEqual(firstcost, 1760, 3)
 
+    def test_skipfailures(self):
+        x = Variable("x")
+        x_min = Variable("x_{min}", [1, 2])
```
Comment (whoburg) [inline gpkit/tests/t_sub.py]: oh cool, I either forgot or was never aware of this shorthand for `("sweep, [1, 2]")`. And I see it's already documented -- nice.

**Pull request merged**